### ISSUE TITLE
Mass "Add Device" via CSV

### ISSUE THREAD

**Issue by joepavitt** [FlowFuse/flowfuse#1369]
### Description

Being able to define and configure device in a csv and upload to the "Devices" page in the UI would save users a lot of time. Columns could include device name, type, project, potentially including device env vars?

### October 2025 Update

The user story remains the same: **As a user managing devices and wanting to rapidly scale in my deployment of remote instances, I want to easily select and pre-populate my device.yaml file for each container, so that I can avoid either the tedious chore of doing this manually or requiring a homemade solution.**

A customer (below) has called attention to this being a blocker to more rapid scaling. See 56:16 in this video: https://fathom.video/calls/440605221

As we are currently focused on scaling FlowFuse, we want to ensure the easiest path possible toward rapid expansion.

For UI work, this could take the form of:

- User navigates to Remote Instates or selects a particular application
- User selects 'Add Instance'
- A fork appears: add a single remote instance, or bulk add instances
- On the Bulk Add path, the user is given the option to select a device.yaml from an existing remote instance, or to provide a new one.
- The user then creates some n new remote instances and assigns them to an application

### Epic/Story

https://github.com/FlowFuse/CloudProject/issues/862

### Requested By

https://app-eu1.hubspot.com/contacts/26586079/record/0-2/164372207852

**Comment by joepavitt** [comment]
**Current Process:**
- 1 by 1, add device and download yml, then export each yml to each device

**My Proposal:**
- Mass Registration, Downloads ymls, and export each yml to each device

It's better than we have, but agree still not ideal

**Comment by gstout52** [comment]
@knolleary Is a version of this functionality already available? https://flowfuse.com/changelog/2025/04/device-provisioning/

**Comment by hardillb** [comment]
@gstout52 that is a different approach to the problem, the token can be used to create any number of devices, but they are all given generic names and only identifiable by the unique id of the network card in the hardware (this doesn't work well for devices in docker containers, which is how the requesting customer works)

The CSV approach allows specific names to be generated for devices that can be more easily mapped to the target device

**Comment by ZJvandeWeg** [comment]
@hardillb When the agent registers, can it dynamically set the name and or other things from the `device.yml`? Is that a cleaner implementation?

**Comment by knolleary** [comment]
The original premise of the provisioning token was to have a single provisioning `device.yml` that could be copied around without the need to per-device customisation. As such, the code doesn't pick up any device-specific properties from that file and auto-generates the name.

It would be a relatively quick win to update the code to allow those properties to be set in the file on a per-device basis. We should validate that approach with the interested customer to see if it would solve the friction for them.

**Comment by gstout52** [comment]
> The original premise of the provisioning token was to have a single provisioning `device.yml` that could be copied around without the need to per-device customisation. As such, the code doesn't pick up any device-specific properties from that file and auto-generates the name.
> 
> It would be a relatively quick win to update the code to allow those properties to be set in the file on a per-device basis. We should validate that approach with the interested customer to see if it would solve the friction for them.

Good idea, @knolleary . After NR Con, please provide this idea in a little more detail and I'll ask the customer for feedback. They describe their current approach at 56:56 here: https://fathom.video/calls/440605221

**Comment by hardillb** [comment]
Implementing this as an update to the device agent to allow the device name to be specified on the cmd line or in the device.yml when using a provisioning token

**Comment by Steve-Mcl** [comment]
@hardillb I have approved (but not merged) [Allow Provisioned devices to be named device-agent#524](https://github.com/FlowFuse/device-agent/pull/524) as it will close this size:5 issue - it is not clear to me if the PR fully satisfies the whole "via CSV" requirements?

**Comment by hardillb** [comment]
@Steve-Mcl https://github.com/FlowFuse/flowfuse/issues/1369#issuecomment-3461009370

**Comment by knolleary** [comment]
Have updated the title to better reflect the functionality shipped - as it no longer involves CSV
